BUGFIX: Ensure users content stream is never left closed after publication

mhsdesign opened the pull request:
Followup to publishing version 3: https://github.com/neos/neos-development-collection/pull/5301

There are 3 cases currently that might lead to a closed content stream if $something went wrong during publication.

### Case 1: Ensure all events are published BEFORE catchup

Otherwise, due to failures in projection or catchup-hooks the process would be immediately interrupted leaving a broken state.

For example a faulty redirect handler hook - that just listens to live events - would be called during publishing. That means the remaining part to publish is already commited and we know we still have work to do to fork the new user content stream and apply the remaining. But the catchup hook would interrupt immediately when the events were catchup'd live. We would be left with a CLOSED user content stream that contains the "same" events that went live during the rebase. Reopening would not help at that point. This is why we must ensure that all events are published BEFORE we do the first catchup.

Further implications:
- running catchup only once should be more performant
- we cannot refetch the current content stream version for what where previously "subcommans" (`forkContentStream`) but we must pass $expectedVersions around from the outside
- we should not run constraint checks after the first `yield` as that would still operate on the old state. Thus all checks are combined above
- closing the content stream will not be catch up'd directly meaning that another process will not get an error because the content stream is not writeable but because due to the version in the database being behind it will get a concurrency exception

### Case 2: Reopen content stream if base workspace was written to during publication

... and a ConcurrencyException is thrown

Introduces a `WorkspacePublicationDuringWritingTest` parallel test (with own cr) to assert that behaviour.

### Case 3: Close content stream a bit later instead of having to reopen in unexpected errors

Previously an error in `extractFromEventStream` because the payload was not correct and yet has to be migrated would lead to a closed content stream which is of course persisted even after fixing the events via migration.

This is still save to do, as the `closeContentStream` will commit the close on the FIRSTly fetched expected version.

Same guarantees, different error behaviour in rare cases.

<!-- 
    Thanks for your contribution, we appreciate it!

    The first section should explain briefly what is changed. 
    Some examples are always nice to showcase the use. 
    The content will be used in the change-logs and addresses 
    developers working with Neos.

    If there are issues regarding the topic of your PR link 
    them here as `related:` or `resolved:`
-->

**Upgrade instructions**

<!-- 
    Add upgrade instructions for breaking changes. 
    Explain who is affected, what has to be adjusted  
-->

**Review instructions**

<!-- 
    If your change is not explained fully by the first section you can 
    add more details here to help the reviewers understand the change.
    We have to understand what you did, why you did it and how we can 
    verify it works correctly and does no harm.
-->

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

mhsdesign committed:
BUGFIX: Ensure all events are published BEFORE catchup

Otherwise, due to failures in projection or catchup-hooks the process would be immediately interrupted leaving a broken state.

For example a faulty redirect handler hook - that just listens to live events - would be called during publishing.
That means the remaining part to publish is already commited and we know we still have work to do to fork the new user content stream and apply the remaining.
But the catchup hook would interrupt immediately when the events were catchup'd live.
We would be left with a CLOSED user content stream that contains the "same" events that went live during the rebase. Reopening would not help at that point.
This is why we must ensure that all events are published BEFORE we do the first catchup.

Further implications:
- running catchup only once should be more performant
- we cannot refetch the current content stream version for what where previously "subcommans" (`forkContentStream`) but we must pass $expectedVersions around from the outside
- we should not run constraint checks after the first `yield` as that would still operate on the old state. Thus all checks are combined above

bwaidelich commented on Neos.ContentRepository.Core/Classes/CommandHandler/CommandBus.php:
```diff
@@ -29,7 +30,10 @@ public function __construct(
+     * The handler only calculate which events they want to have published,
+     * but do not do the publishing themselves
+     *
+     * @return EventsToPublish|YieldedEventsToPublish
```
Is that really easier to read than
```suggestion
     * @return EventsToPublish|\Generator<EventsToPublish>
```
?

mhsdesign replied:
adjusted :) (without complaining because i did all the argumentation myself in my head ... over dayyyyys :D)

bwaidelich commented on Neos.ContentRepository.Core/Classes/CommandHandler/CommandHandlerInterface.php:
```diff
@@ -23,7 +25,7 @@ public function canHandle(CommandInterface $command): bool;
      * it's allowed to yield the events to interact with the control flow of event publishing.
      *
-     * @return EventsToPublish|\Generator<int, EventsToPublish>
+     * @return EventsToPublish|YieldedEventsToPublish
```
same here, I think the custom type does more harm than it helps to be honest.
(but not a big issue of course)

bwaidelich commented on Neos.ContentRepository.Core/Classes/EventStore/EventPersister.php:
```diff
@@ -31,15 +32,22 @@ public function publishEvents(ContentRepository $contentRepository, EventsToPubl
+    /**
+     * @throws ConcurrencyException in case the expectedVersion does not match
+     */
+    public function publishWithoutCatchup(EventsToPublish $eventsToPublish): CommitResult
```
This makes the `EventPersister` even more ugly than it already is :)
But I am confident that we'll be able to simplify the whole thing with #5321

bwaidelich commented on Neos.ContentRepository.Core/Classes/Feature/ContentStreamHandling.php:
```diff
@@ -67,29 +37,26 @@ private function closeContentStream(
      * @phpstan-pure this method is pure, to persist the events they must be handled outside
      */
-    private function reopenContentStream(
+    private function reopenContentStreamWithoutConstraints(
```
I had suggested to rename this method but only if there's another version that _does_ constraint checks.
In this case I would just keep the previous name (and the new inline comment) since it's an implementation detail.
If you want to keep the new name it should be "...WithoutConstraintChecks"

bwaidelich commented on Neos.ContentRepository.Core/Classes/Feature/ContentStreamHandling.php:
```diff
@@ -133,25 +91,19 @@ private function forkContentStream(
      * @phpstan-pure this method is pure, to persist the events they must be handled outside
      */
-    private function removeContentStream(
+    private function removeContentStreamWithoutConstraints(
```
Same here: I would only add the suffix if there's a version _with_ constraint checks

bwaidelich reviewed (commented):
I think it's a great start and the LoC balance is negative even:
![image](https://github.com/user-attachments/assets/9efc2fd3-232e-46f9-87d5-14abfa649c47)

kitsunet commented on Neos.ContentRepository.Core/Classes/ContentRepository.php:
```diff
@@ -94,19 +94,35 @@ public function __construct(
+
+        // control-flow aware command handling via generator
+        try {
+            $yieldedEventsToPublish = $toPublish->current();
```
The question is why would we then even need to yield, because we might as well return one EventsToPublish then, if there is no catch up in between?! Also means e.g. pre-emptive closing of CS is pointless as other processes constraint checks will never see it when it would be useful to have.

mhsdesign replied:
> pre-emptive closing of CS is pointless as other processes constraint checks will never see it when it would be useful to have.

we thought about that:

> > closing the content stream will not be catch up'd directly meaning that another process will not get an error because the content stream is not writeable but because due to the version in the database being behind it will get a concurrency exception


thats why we yield. Further - still todo - we need to reopen the user workspace for the unlikely case that the matching events could not be published, as we dont lock or close the base workspace ... maybe we should though ...

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into bugfix/publishing-ensure-contentstream-not-closed

mhsdesign committed:
BUGFIX: Fix reopen content stream if base workspace was written to during publication

... and a ConcurrencyException is thrown

Introduces a `WorkspacePublicationDuringWritingTest` parallel test (with own cr) to assert that behaviour.

mhsdesign committed:
TASK: Fix parallel tests by ensuring only one is run at time

That allows us to use the same content repository. Previously a super slow paratest would lead that another testcase will already be started and its setup then run twice at the end.

https://github.com/paratestphp/paratest/discussions/905

mhsdesign committed:
TASK: Simplify code and remove reopen cs logic into `publishWorkspace`

mhsdesign committed:
TASK: Only fetch content stream once for constraint checks

mhsdesign committed:
TASK: Adjust .composer json

mhsdesign committed:
TASK: Do not send `$commitResult` to generator but calculate expected version instead

Also readd lost documentation and simplifies the `handle`

The ->throw logic was initially introduced via

https://github.com/neos/neos-development-collection/pull/5315

but then removed again as we thought it was no longer needed.

mhsdesign committed:
TASK: Inline now simplified `YieldedEventsToPublish` virtual type again

mhsdesign committed:
TASK: Wrap rebaseable command extraction into `finally` block to ensure content stream is never left closed

During the beta phase it can happen that user forget to apply a migration to migrate the stored commands in the even metadata, upon publish this would close the content stream and fail directly afterward.

Applying the migration then would not be enough as the content stream is a closed state and has to be repaired manually.

Event thought this is not super likely, its not unlikely as well and the case during publication were we rely on things that might not be that way.

As an alternative we could discuss doing the closing after acquiring the rebaseable commands.

mhsdesign committed:
TASK: Close content stream a bit later instead of having to reopen in many edge cases

Alternative fix for d27f83f024c703c4ef3b55f315fea73fb78b5dd9

Previously an error in `extractFromEventStream` because the payload was not correct and yet has to be migrated would lead to a closed content stream which is of course persisted even after fixing the events via migration.

This is still save to do, as the `closeContentStream` will commit the close on the FIRSTly fetched expected version.

Same guarantees, different error behaviour in rare cases.

mhsdesign committed:
TASK: Add proper docs to `EventPersister`

;)

mhsdesign committed:
TASK: Adjust naming of `removeContentStreamWithoutConstraintChecks`

;)

bwaidelich commented on .composer.json:
```diff
@@ -25,8 +25,9 @@
+    "test:paratest-cli": "../../bin/paratest --debug -v --functional --processes 2 --colors --stop-on-failure -c ../../Build/BuildEssentials/PhpUnit/FunctionalTests.xml",
     "test:parallel": [
-      "FLOW_CONTEXT=Testing/Behat ../../bin/paratest --debug -v --functional --group parallel --processes 2 --colors --stop-on-failure -c ../../Build/BuildEssentials/PhpUnit/FunctionalTests.xml Neos.ContentRepository.BehavioralTests/Tests/Functional/Feature/WorkspacePublication/WorkspaceWritingDuringPublication.php"
+      "for f in Neos.ContentRepository.BehavioralTests/Tests/Parallel/**/*Test.php; do composer test:paratest-cli $f; done"
```
(how) is that related and what does it do?

mhsdesign replied:
another paratest so we need to execute all not only one. that was more tricky than i wanted it to be:

see

https://github.com/neos/neos-development-collection/pull/5342/commits/18072f695823924fae05c2455a8b356fab8a96a2

and my question here:

https://github.com/paratestphp/paratest/discussions/905

bwaidelich commented on Neos.ContentRepository.BehavioralTests/Tests/Parallel/AbstractParallelTestCase.php:
```diff
@@ -52,14 +52,15 @@ final protected function awaitFile(string $filename): void
-                throw new \Exception('timeout while waiting on shared lock');
+            clearstatcache(true, $filename);
+            if ($waiting > 60000) {
+                throw new \Exception('timeout while waiting on file ' . $filename);
```
Should we maybe extract those changes to the infra/test setup to a separate pr? ( mostly to allow to get a useful log via git blame)

mhsdesign replied:
nein :D. We already have way to many prs and discussions flying around. 
I did fix the previous para test with this dedicated pr but imo its super fine to push a few minor fixes on it :)
https://github.com/neos/neos-development-collection/pull/5344

and i used dedicated commits ;)

bwaidelich replied:
"nein" is not really an appropriate response to a suggestion in a code review 😆 
But I get your point, and thanks for the dedicated commits

bwaidelich commented on Neos.Neos/Classes/Domain/Model/UserId.php:
```diff
@@ -15,7 +15,7 @@ public function __construct(
     ) {
         if (!preg_match('/^([a-z0-9\-]{1,40})$/', $value)) {
-            throw new \InvalidArgumentException(sprintf('Invalid user id "%s" (a user id must only contain lowercase characters, numbers and the "-" sign).', 1718293224));
+            throw new \InvalidArgumentException(sprintf('Invalid user id "%s" (a user id must only contain lowercase characters, numbers and the "-" sign).', $this->value), 1718293224);
```
Oops. Good catch

bwaidelich commented on Neos.ContentRepository.Core/Classes/ContentRepository.php:
```diff
@@ -94,19 +95,48 @@ public function __construct(
+            $eventsToPublish = $this->enrichEventsToPublishWithMetadata($toPublish);
+            if ($eventsToPublish->events->isEmpty()) {
+                return;
+            }
```
This check could be moved before the enrichment I suppose
 But it's not important and it will hopefully be removed with my other pr anyways

mhsdesign replied:
adjusted :)

bwaidelich commented on Neos.ContentRepository.Core/Classes/ContentRepository.php:
```diff
@@ -94,19 +95,48 @@ public function __construct(
+            if ($eventsToPublish->events->isEmpty()) {
+                return;
+            }
+            $this->eventPersister->publishWithoutKetchup($eventsToPublish);
```
Did you really call it Ketchup to troll us? 🙃

mhsdesign replied:
Maybe for jon :D 
It will be removed either way with the subscription engine:

> TODO Will be refactored via https://github.com/neos/neos-development-collection/pull/5321

:D

mhsdesign replied:
adjusted as requested :) https://github.com/neos/neos-development-collection/pull/5342/commits/e12c6414aa5a89bc5835f002ef69abab4b12d76e

bwaidelich commented on Neos.ContentRepository.Core/Classes/EventStore/EventPersister.php:
```diff
@@ -24,22 +25,31 @@ public function __construct(
+     * TODO Will be refactored via https://github.com/neos/neos-development-collection/pull/5321
+     * @throws ConcurrencyException in case the expectedVersion does not match
+     */
+    public function publishWithoutKetchup(EventsToPublish $eventsToPublish): CommitResult
```
🍅

kitsunet replied:
refactored or not, that other PR is still Draft, and while I have full trust in bwaidelich, I don't like to bet on anything here, we should release, maybe it must be without subscription engine 🤷 who knows, but I don't really awnt ot releaes with ketchup. So please rename :)

dlubitz replied:
Thanks for pointing out how careful I did my review 🙈

dlubitz replied:
I guess I had 🍅 🍅 on my eyes.

mhsdesign replied:
i renamed it again to the proper naming in https://github.com/neos/neos-development-collection/pull/5342/commits/dfd457360a13aee60ae9c22d9a19c5c4b5683dcb :D I had it corrected but was tempted as we had this fun sprint today and what would be a sprint day without Schabernack :D

bwaidelich commented on Neos.ContentRepository.Core/Classes/Feature/Common/ConstraintChecks.php:
```diff
@@ -81,14 +81,9 @@ protected function requireContentStream(
-                1521386692
-            );
-        }
+        $closedState = $commandHandlingDependencies->isContentStreamClosed($contentStreamId);
```
Nitpick: rename to `$contentStreamIsClosed`

bwaidelich reviewed (commented):
Some, probably semi-helpful, comments. Looks great otherwise

mhsdesign committed:
TASK: Improve assertions of WorkspacePublicationDuringWritingTest

mhsdesign committed:
TASK: Assert that in WorkspaceWritingDuringRebaseTest that the workspace is still the original content stream

mhsdesign committed:
TASK: Naming things and suggestion from code review :)

bwaidelich reviewed (approved):
+1 by reading and running the tests

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into bugfix/publishing-ensure-contentstream-not-closed

dlubitz reviewed (commented):
Look good by reading. But as a disclaimer ... I'm not that deep into this topic to fully understand. 🤷‍♂️

mhsdesign committed:
TASK: Fix tests after bastis command test overhaul

mhsdesign committed:
TASK: Recorrect naming of method again (my code editor seems to misbehave :))
